File: web/src/lib/tasks.ts
import type { TaskListQuery, TaskRunSummary, TaskStreamEnvelope, TaskSummary } from "@/lib/types";

export function buildTaskSearchParams(query: TaskListQuery) {
  const params = new URLSearchParams();

  if (query.profile_id) {
    params.set("profile_id", query.profile_id);
  }
  if (query.kind) {
    params.set("kind", query.kind);
  }
  if (query.status) {
    params.set("status", query.status);
  }
  if (query.trigger) {
    params.set("trigger", query.trigger);
  }
  if (query.running_only) {
    params.set("running_only", "true");
  }
  if (query.since != null) {
    params.set("since", String(query.since));
  }
  if (query.limit != null) {
    params.set("limit", String(query.limit));
  }
  if (query.cursor) {
    params.set("cursor", query.cursor);
  }

  return params;
}

export function matchesTaskQuery(run: TaskRunSummary, query: TaskListQuery) {
  if (query.profile_id && run.profile_id !== query.profile_id) {
    return false;
  }
  if (query.kind && run.kind !== query.kind) {
    return false;
  }
  if (query.status && run.status !== query.status) {
    return false;
  }
  if (query.trigger && run.trigger !== query.trigger) {
    return false;
  }
  if (query.running_only && run.status !== "running") {
    return false;
  }
  if (query.since != null && run.created_at < query.since) {
    return false;
  }
  return true;
}

export function sortTaskRuns(runs: TaskRunSummary[]) {
  return [...runs].sort((left, right) => {
    if (right.created_at !== left.created_at) {
      return right.created_at - left.created_at;
    }
    return right.run_id.localeCompare(left.run_id);
  });
}

export function summarizeTaskRuns(runs: TaskRunSummary[]): TaskSummary {
  const summary: TaskSummary = {
    total_runs: 0,
    queued_runs: 0,
    running_runs: 0,
    failed_runs: 0,
    succeeded_runs: 0,
    skipped_runs: 0,
    last_run_at: null,
  };

  for (const run of runs) {
    summary.total_runs += 1;
    if (run.status === "queued") {
      summary.queued_runs += 1;
    }
    if (run.status === "running") {
      summary.running_runs += 1;
    }
    if (run.status === "failed") {
      summary.failed_runs += 1;
    }
    if (run.status === "succeeded") {
      summary.succeeded_runs += 1;
    }
    if (run.status === "skipped") {
      summary.skipped_runs += 1;
    }

    const candidate = run.finished_at ?? run.started_at ?? run.created_at;
    summary.last_run_at = Math.max(summary.last_run_at ?? 0, candidate);
  }

  return summary;
}

export function parseTaskEnvelope<T>(payload: string): TaskStreamEnvelope<T> {
  return JSON.parse(payload) as TaskStreamEnvelope<T>;
}

File: src/tasks.rs
use crate::models::{
    TaskListQuery, TaskRunDetail, TaskRunEventRecord, TaskRunStatus, TaskRunSummary, TaskSummary,
};

#[derive(Debug, Clone)]
pub enum TaskBusEvent {
    RunUpsert(TaskRunSummary),
    RunEvent(TaskRunEventRecord),
}

pub fn matches_task_query(run: &TaskRunSummary, query: &TaskListQuery) -> bool {
    if let Some(profile_id) = &query.profile_id
        && profile_id != "all"
        && &run.profile_id != profile_id
    {
        return false;
    }
    if let Some(kind) = query.kind
        && run.kind != kind
    {
        return false;
    }
    if let Some(status) = query.status
        && run.status != status
    {
        return false;
    }
    if let Some(trigger) = query.trigger
        && run.trigger != trigger
    {
        return false;
    }
    if query.running_only && run.status != TaskRunStatus::Running {
        return false;
    }
    true
}

pub fn build_task_summary(runs: &[TaskRunSummary]) -> TaskSummary {
    let mut summary = TaskSummary::default();
    for run in runs {
        summary.total_runs += 1;
        match run.status {
            TaskRunStatus::Queued => summary.queued_runs += 1,
            TaskRunStatus::Running => summary.running_runs += 1,
            TaskRunStatus::Failed => summary.failed_runs += 1,
            TaskRunStatus::Succeeded => summary.succeeded_runs += 1,
            TaskRunStatus::Skipped => summary.skipped_runs += 1,
        }

        let candidate = run.finished_at.or(run.started_at).unwrap_or(run.created_at);
        summary.last_run_at = Some(
            summary
                .last_run_at
                .map_or(candidate, |current| current.max(candidate)),
        );
    }
    summary
}

pub fn to_detail(run: TaskRunSummary, events: Vec<TaskRunEventRecord>) -> TaskRunDetail {
    TaskRunDetail {
        run,
        events: events.into_iter().map(|event| event.as_public()).collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{TaskRunKind, TaskRunStage, TaskRunTrigger};

    fn sample_run(profile_id: &str, kind: TaskRunKind, status: TaskRunStatus) -> TaskRunSummary {
        TaskRunSummary {
            run_id: format!("run-{profile_id}"),
            profile_id: profile_id.to_string(),
            kind,
            trigger: TaskRunTrigger::Schedule,
            status,
            stage: TaskRunStage::Queued,
            progress_current: Some(0),
            progress_total: None,
            created_at: 1,
            started_at: None,
            finished_at: None,
            summary_json: None,
            error_code: None,
            error_message: None,
        }
    }

    #[test]
    fn matches_task_query_treats_all_profile_as_aggregate_scope() {
        let run = sample_run(
            "edge-jp",
            TaskRunKind::MetadataRefreshFull,
            TaskRunStatus::Succeeded,
        );
        let query = TaskListQuery {
            profile_id: Some("all".to_string()),
            ..TaskListQuery::default()
        };
        assert!(matches_task_query(&run, &query));
    }

    #[test]
    fn matches_task_query_applies_non_profile_filters() {
        let run = sample_run(
            "default",
            TaskRunKind::SubscriptionSync,
            TaskRunStatus::Running,
        );
        let query = TaskListQuery {
            kind: Some(TaskRunKind::MetadataRefreshIncremental),
            ..TaskListQuery::default()
        };
        assert!(!matches_task_query(&run, &query));
    }
}
